### ISSUE TITLE
puddletag spawns invisible ghost windows in bspwm

### ISSUE THREAD

**Issue by k0mmsussert0d** [puddletag/puddletag#648]
Please see the video attached: https://streamja.com/wkqG1

I'm using a puddletag alongside with my terminal. Desired layout is puddletag occupying left half screen with the terminal occupying right half screen. However, puddletag has already spawned some transparent windows that cannot be killed with standard `super+q` combo (`bspc node -k`). You can see, how I switch puddletag with those ghost windows using `super+shift+h,j,k,l` (`bspc node -s {west,south,north,east}` to visualize their size. At this point, screen layout is somewhat like:
```
--------------------
| P | X |     T    |
|   X   |          |
--------------------

P - puddletag
T - terminal
X - ghost window
```

Then, I go to monocle layout (`super + m` a.k.a. `bspc desktop -l next`) to use puddletag a little bit. After exiting the monocle layout, yet another ghost window is spawned in top left quarter of the screen. Now, moving puddletag window around reveals layout to resemble:
```
------------------------
| P | X | X |     T    |
|      X    |          |
------------------------
```

Windows remain after puddletag is closed. I found no way to get rid of them other than restarting X11.
Inspecting those areas with `xprop` provides no hint. It's as if a desktop (wallpaper) area got clicked.

*Versions:*
```
$ pacman -Q --info puddletag
Name            : puddletag
Version         : 2.0.1-2

$ pacman -Q --info xorg-server
Name            : xorg-server
Version         : 1.20.13-3

 $ pacman -Q --info bspwm
Name            : bspwm
Version         : 0.9.10-1
```

Running on Arch Linux with 5.14.15 kernel. Let me know if you need any other details. Thanks!

EDIT: Just discovered one can get rid of ghost windows with `bspc node -c`

**Comment by corubba** [comment]
I am not going to lie, this was a hard nut to crack. We are talking solid tungsten here. But no matter how small the hammer, if you chisel long enough you will crack it.

tl;dr I prepared a [bugfix](https://github.com/corubba/puddletag/tree/bugfix/ghost-window) that solved the issue, at least as far as my testing showed. Please give it a try and report if it also fixes the ghost window issue for you.

So here's what I found out: When puddletag shows the little window/popup/modal with the progressbar (for example when opening a folder), it will create/show a new window and immediately close/destroy it, followed by showing the actual progress window. This intermediate window is caused by an "imperfection" (some may be inclined to call it a bug) in puddletag (more further below). This window only exists for a splitsecond, and because of that bspwm somehow misses the window-closed event and keeps the window open while Xorg already fully closed it. This can easily be checked by looking what window ids bspwm (`bspc query -N -n '.window'`) and Xorg (`xwininfo -tree -root`) know about, the ids of the ghost windows will only be reported by bspwm which is also what you saw. This may even be just a race condition in bspwm, because if you keep that intermediate window open long enough (for example by inserting a `time.sleep(1)` [here](https://github.com/puddletag/puddletag/blob/a0d4ab4b22d04575fcf974cb1086293468ccf6d4/puddlestuff/puddleobjects.py#L2208)) you can see the intermediate window appear and also properly disappear.
Instead of investigating further into bspwm, I turned my intention to why that intermediate only-shown-for-a-splitsecond window appears in the first place, and how to get rid of it. During construction of the progress window, we create a layout without a parent (so it is a root node) and attach the Cancel button to it. Depending on whether the cancel button should be shown, its visibility will be explicitly set to `true` or `false`. Explicitly setting the visibility to `true` will trigger a check to see if it needs to be shown, which goes up the node tree to see if any parents a hidden. In our case that tree only has two nodes (the button and the layout) and both are visible (the layout by default, and the button explicitly). This triggers Qt to create a new window (because there is none associated yet with that tree) and show layout+button, which is that intermediate window we see. A few code-lines later, the layout is attached to another layout (`vbox`) which in turn gets attached to the ProgressWin QDialog. When this "reparenting" happens, Qt will find that window to contain nothing anymore, and close/destroy it. And because the button now has the hidden QDialog as a parent, it won't be shown either.
To remove the intermediate window, you can either move the `cancel.setVisibility(boolean)` to after the reparenting happened, or just not set the button explicitly to visible (which is the way I have done it).

**Comment by k0mmsussert0d** [comment]
Hi @corubba 

First things first, I want to express my deepest gratitude towards you for all the efforts you've put into tracking down the cause of the issue and eventually resolving it. To be honest, at the moment of submitting this issue I've had already suspected it will never be resolved. Mainly due to very specific set of predicates required for it to occur and, as a result, potentially near-to-zero volume of users affected by it. I have also tried to investigate the issue on my own, but having virtually no knowledge of Qt or building GUIs in Python in general, I was unable to come up with any relevant conclusion (although now at least I can see that I was looking at the right place, lol). Using bspwm is a very specific and uncommon factor required to reproduce the bug, which makes me extend my gratitude towards your person even more.

While testing your fork, I tried out different actions displaying progress bar window and could not observe ghost windows spawning anymore. I've noticed a different discrepancy, although I'm unsure of it being caused by your fix. Namely, directories in *Filesystem* tab that have no subdirectories, still display drop-down arrow next to them. I could not observe the same behavior in the newest `2.1.1` version, so I thought there might be some regression. Other than that, the issue that has been bringing me tons of problems and frustration for the last year is gone and that makes me more than happy :D 

![Screenshot_20220313_174713](https://user-images.githubusercontent.com/26551643/158070998-5256ac47-991c-4658-b404-7414013a20d2.png)
